### ISSUE TITLE
Upgrade fails due to dependencies on PCRE

### ISSUE THREAD

**Issue by adaminfinitum** [Homebrew/brew#2396]
Every time I have tried to update/upgrade brew in the last month or so it has failed due to dependencies on PCRE.

```
$ brew config
HOMEBREW_VERSION: 1.1.11-159-g422afa0b49
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 422afa0b49bae84e707cb3f66faf2da32cb4c462
Last commit: 10 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 6e40e10afe4e85b56ab0e2bd8e0ddc20fce47905
Core tap last commit: 71 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: octa-core 64-bit haswell
Homebrew Ruby: 2.0.0-p481
Clang: 7.0 build 700
Git: 2.12.2 => /usr/local/bin/git
Perl: /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
Java: 1.8.0_121, 1.8.0_51, 1.6.0_65-b14-466.1
macOS: 10.10.5-x86_64
Xcode: 7.2.1
CLT: 7.2.0.0.1.1447826929
X11: 2.7.11 => /opt/X11
```

```
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!

Warning: Python is installed at /Library/Frameworks/Python.framework

Homebrew only supports building against the System-provided Python or a
brewed Python. In particular, Pythons installed to /Library can interfere
with other software installs.

Error: pcre: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:4: syntax error, unexpected <<, expecting keyword_end
<<<<<<< Updated upstream
  ^
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:8: syntax error, unexpected ===, expecting keyword_end
=======
   ^
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:12: syntax error, unexpected >>, expecting keyword_end
>>>>>>> Stashed changes
  ^
```
I'm pretty sure the Python warning isn't part of this, it existed long before this issue arose.

I have tried lots of things (update, upgrade, reinstall, uses, cleanup, prune, repair, missing, postinstall, reinstall etc.). I know that all of those are relevant but the problem manifested after I migrated my apps to casks.

I also made a backup of `pcre.rb` and deleted the original, etc.

I kept getting messages about `atk` requiring `pcre` so I tried uninstalling `atk` but that didn't help, I just kept getting messages about how `gtk+` required `pcre`.

I just ran through this series of steps without luck:
```
/usr/local/Homebrew
$ brew search pcre
pcre                                                        pcre++                                                      pcre2 ✔

/usr/local/Homebrew 
$ brew update
Already up-to-date.

/usr/local/Homebrew
$ brew upgrade
Upgrading 18 outdated packages, with result:
glib 2.52.0, gobject-introspection 1.52.0, harfbuzz 1.4.5, imagemagick 7.0.5-4, libgpg-error 1.27, mercurial 4.1.1, node 7.7.4, numpy 1.12.1, homebrew/dupes/openssh 7.5p1, homebrew/php/php56 5.6.30_6, pkg-config 0.29.2, pyenv 1.0.9, python3 3.6.1, shared-mime-info 1.8_1, sqlmap 1.1.3, tcpreplay 4.2.1, vim 8.0.0512, wireshark 2.2.5
Error: pcre: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:4: syntax error, unexpected <<, expecting keyword_end
<<<<<<< Updated upstream
  ^
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:8: syntax error, unexpected ===, expecting keyword_end
=======
   ^
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:12: syntax error, unexpected >>, expecting keyword_end
>>>>>>> Stashed changes
  ^
```

Here's what I get when I try to install:
```
$ brew install pcre
Error: pcre: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:4: syntax error, unexpected <<, expecting keyword_end
<<<<<<< Updated upstream
  ^
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:8: syntax error, unexpected ===, expecting keyword_end
=======
   ^
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb:12: syntax error, unexpected >>, expecting keyword_end
>>>>>>> Stashed changes
  ^
==> Searching for similarly named formulae...
These similarly named formulae were found:
pcre                                                        pcre++                                                      pcre2 ✔
To install one of them, run (for example):
  brew install pcre
==> Searching taps...
Error: No formulae found in taps.
```

I don't know why it's suddenly returning `pcre` when I search (maybe because I restored `pcre.rb` from the backup copy I made before deleting it) , it wasn't before but it won't let me install it and still isn't working.

**Comment by axiac** [comment]
This is not a Homebrew issue. The lines that start with `>>>>>>>`, `=======` and `<<<<<<<` in a file of a Git repository are the sign of an unsolved conflict. You either did some changes to the file `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pcre.rb` or, unintentionally, ran some Git commands (merge, rebase etc) in the tap's directory.

I assume you didn't change any of the tap files on purpose (on the contrary I think you should know how to handle the situation).

Try to fix the tap by running:

    $ cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
    $ git fetch origin
    $ git checkout master
    $ git reset --hard origin/master

After that run `brew doctor`. The warning about Python should still be there but the errors about pcre.rb shouldn't be any more.

**Comment by MikeMcQuaid** [comment]
@axiac A simpler way that will run that is `brew update-reset` but: thanks ❤️.

**Comment by adaminfinitum** [comment]
Thanks for your help, it worked; I tried `brew update-reset` first and it fixed the problem.

I know brew uses git (and I understand the steps outlined by @axiac) but I'm clearly missing something...

Is there something important about `cd`ing into the relevant directory and using git directly?

Maybe the `brew` command can't update or reset things in the homebrew core?

I'd appreciate an explanation of how/why you knew this was the problem and that these steps would fix it.

Thanks.

**Comment by MikeMcQuaid** [comment]
> I know brew uses git (and I understand the steps outlined by @axiac) but I'm clearly missing something...

Homebrew doesn't resolve Git merge conflicts for you, you have to do that. `brew update-reset` will resolve them but blow away all changes in the process.

**Comment by axiac** [comment]
@adaminfinitum A possible scenario (it can be completely wrong as well): for some reason you `cd` into the Homebrew's directory. Curiosity is a very good reason. An error reported by `brew update` can be another good reason as well. You changed one file. The next `brew update` discovered the local branch diverged from the remote branch. It tried to preserve your local changes (stashed them, pulled the recent commits from the remote repository, tried to unstash the changes). Unfortunately the operation failed because conflicting changes were operated on the two sides. The file ended up containing both versions of the changed code, guarded by the conflict markers (`<<<<<<<`, `=========` and `>>>>>>>`). Any further update was stuck until you resolve the conflicts and commit. Resolving the conflict never happened.

If you know how to use Git I hope the explanation above give you an idea about what happened. Another scenario is possible as well. If you don't use Git then probably the paragraph above is just programmers gibberish and doesn't make much sense. In this case just take my word and stay away from the Homebrew directory :-)

`brew update-reset` runs Git commands similar to those I posted in the previous comment on all local Git repositories used by Homebrew (code and taps). It throws any local changes and gets the local repos back in sync with the official Homebrew repos, putting back on the track the instances of `brew` that got stuck in situations like yours.

**Comment by vitorgalvao** [comment]
Quick question: why not document `update-reset`? It seems like it’s in neither the `man` page nor any help documents.

**Comment by MikeMcQuaid** [comment]
> Quick question: why not document update-reset? It seems like it’s in neither the man page nor any help documents.

Basically because it's extremely destructive. It's useful just for a command we can tell people to run but don't want people to run it themselves. Note that only people who have run developer commands e.g. `brew audit` will end up in this state as we've assumed they know how to use Git and fix those errors.

**Comment by adaminfinitum** [comment]
@axiac I've been playing with / learning Git for a couple years but I don't work as a developer so I'm not fluent. I understand the basics but (for example) I struggle to know exactly how and when to use `stash` vs. `merge`.

In this case, because I was getting messages that PCRE has been removed from Homebrew, I was wondering if I just needed to do a recursive find and replace in Homebrew to substitute PCRE2 for PCRE...but I have never had to do something like that with Homebrew before (and it just seemed like a bad idea that would probably cause more problems I wouldn't know how to fix).

As you noted, I stay away from the Homebrew directory and rely on `brew`...my compliments to the developers, when something is wrong, `brew` has always told me exactly how to fix it, until now.

Anyway, I learn best by doing so thanks everybody for your help, patience and explanations.
